File: packages/schemas/ts/src/knowledge.ts
import { z } from "zod";
import { SchemaVersionV1 } from "./schema-version";

/**
 * Knowledge-layer domain records (open records: additive fields from newer
 * minors are ignored). The chain is strict: Source → Evidence → Claim →
 * Knowledge; claims and evidence are never folded into node summaries.
 */
export const SourceKind = z.enum(["web", "paper", "book", "github", "dataset", "news", "other"]);

export const QualityRating = z.enum(["high", "medium", "low", "unevaluated"]);

export const SourceQualitySchema = z
  .object({
    authority: QualityRating.optional(),
    fitness: QualityRating.optional(),
    evaluated_at: z.string().optional(),
    notes: z.string().optional(),
  })
  .strict();
export type SourceQuality = z.infer<typeof SourceQualitySchema>;

export const SourceSchema = z.object({
  schema_version: SchemaVersionV1,
  source_id: z.string().min(1),
  project_id: z.string().min(1),
  url: z.string().min(1),
  title: z.string().min(1),
  kind: SourceKind,
  published_at: z.string().optional(),
  retrieved_at: z.string(),
  languages: z.array(z.string().min(2)).optional(),
  dedup_key: z.string().min(1),
  quality: SourceQualitySchema.optional(),
});
export type Source = z.infer<typeof SourceSchema>;

export const ContentFormat = z.enum(["html", "pdf", "markdown", "plain", "json"]);

export const SourceContentSchema = z.object({
  schema_version: SchemaVersionV1,
  content_id: z.string().min(1),
  source_id: z.string().min(1),
  format: ContentFormat,
  content_hash: z.string().min(8),
  language: z.string().optional(),
  fetched_at: z.string(),
  cache_ref: z.string().optional(),
});
export type SourceContent = z.infer<typeof SourceContentSchema>;

export const KnowledgeNodeType = z.enum([
  "concept",
  "person",
  "organization",
  "company",
  "paper",
  "book",
  "experiment",
  "event",
  "technology",
  "product",
  "application",
  "policy",
  "dataset",
  "controversy",
]);

export const NodeStatus = z.enum(["active", "merged", "superseded"]);

export const Confidence = z.enum([
  "confirmed",
  "high",
  "medium",
  "low",
  "unverified",
  "conflicting",
]);

export const KnowledgeNodeSchema = z.object({
  schema_version: SchemaVersionV1,
  node_id: z.string().min(1),
  project_id: z.string().min(1),
  type: KnowledgeNodeType,
  title: z.string().min(1),
  aliases: z.array(z.string().min(1)).optional(),
  summary: z.string().optional(),
  status: NodeStatus.optional(),
  confidence: Confidence,
  source_ids: z.array(z.string().min(1)).optional(),
  claim_ids: z.array(z.string().min(1)).optional(),
  metadata: z.record(z.unknown()).optional(),
  created_at: z.string(),
  updated_at: z.string(),
});
export type KnowledgeNode = z.infer<typeof KnowledgeNodeSchema>;

export const ClaimProvenanceSchema = z
  .object({
    created_by: z.enum(["user", "ai"]).optional(),
    prompt_id: z.string().optional(),
    prompt_version: z.string().optional(),
    task_id: z.string().optional(),
  })
  .strict();
export type ClaimProvenance = z.infer<typeof ClaimProvenanceSchema>;

export const ReviewState = z.enum(["unreviewed", "needs-review", "reviewed"]);

/**
 * Released minors of the claim contract (ADR-016): 1.1 splits the formerly
 * conflated `status` into `status` (review lifecycle) + `confidence`.
 */
export const ClaimSchemaVersion = z.enum(["1.0", "1.1"]);
export type ClaimSchemaVersion = z.infer<typeof ClaimSchemaVersion>;

/** Review lifecycle of a claim (ADR-016); evidence strength lives in `confidence`. */
export const ClaimStatus = z.enum(["draft", "needs_review", "confirmed", "superseded"]);
export type ClaimStatus = z.infer<typeof ClaimStatus>;

export const ClaimSchema = z.object({
  schema_version: ClaimSchemaVersion,
  claim_id: z.string().min(1),
  project_id: z.string().min(1),
  subject_node_id: z.string().min(1),
  predicate: z.string().min(1),
  object: z.string().min(1),
  scope: z.string().optional(),
  status: ClaimStatus,
  confidence: Confidence.default("unverified"),
  evidence_ids: z.array(z.string().min(1)).optional(),
  provenance: ClaimProvenanceSchema.optional(),
  review_state: ReviewState.optional(),
  created_at: z.string(),
  updated_at: z.string(),
});
export type Claim = z.infer<typeof ClaimSchema>;

export const EvidenceDirection = z.enum(["support", "contradict"]);

export const EvidenceLocatorSchema = z
  .object({
    quote: z.string().optional(),
    page: z.string().optional(),
    section: z.string().optional(),
    fragment: z.string().optional(),
  })
  .strict();
export type EvidenceLocator = z.infer<typeof EvidenceLocatorSchema>;

export const EvidenceSchema = z.object({
  schema_version: SchemaVersionV1,
  evidence_id: z.string().min(1),
  claim_id: z.string().min(1),
  source_id: z.string().min(1),
  direction: EvidenceDirection,
  locator: EvidenceLocatorSchema,
  extraction_method: z.enum(["ai", "user"]).optional(),
  retrieved_at: z.string(),
  revised_at: z.string().optional(),
  superseded_by: z.string().optional(),
});
export type Evidence = z.infer<typeof EvidenceSchema>;

export const RelationConfidence = z.enum(["confirmed", "high", "medium", "low", "unverified"]);

export const RelationSchema = z.object({
  schema_version: SchemaVersionV1,
  relation_id: z.string().min(1),
  from_node_id: z.string().min(1),
  to_node_id: z.string().min(1),
  predicate: z.string().min(1),
  direction: z.enum(["directed", "undirected"]).optional(),
  confidence: RelationConfidence,
  status: NodeStatus.optional(),
  source_ids: z.array(z.string().min(1)).optional(),
  claim_ids: z.array(z.string().min(1)).optional(),
  provenance: ClaimProvenanceSchema.optional(),
  created_at: z.string(),
  updated_at: z.string().optional(),
});
export type Relation = z.infer<typeof RelationSchema>;

export const ArtifactKind = z.enum([
  "vault-export",
  "report",
  "graph-snapshot",
  "coverage-report",
  "other",
]);

export const ArtifactSchema = z.object({
  schema_version: SchemaVersionV1,
  artifact_id: z.string().min(1),
  project_id: z.string().min(1),
  kind: ArtifactKind,
  run_id: z.string().optional(),
  path: z.string().optional(),
  format: z.enum(["markdown", "json", "png", "svg"]).optional(),
  note: z.string().optional(),
  created_at: z.string(),
});
export type Artifact = z.infer<typeof ArtifactSchema>;

File: packages/schemas/rust/src/knowledge.rs
//! Knowledge-layer domain record bindings (open records).
//!
//! Chain: Source -> Evidence -> Claim -> Knowledge. Claims and evidence are
//! never folded into node summaries; conflicting claims coexist.

use serde::{Deserialize, Serialize};

use crate::schema_version::SchemaVersionV1;

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum SourceKind {
    Web,
    Paper,
    Book,
    Github,
    Dataset,
    News,
    Other,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum QualityRating {
    High,
    Medium,
    Low,
    Unevaluated,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum ContentFormat {
    Html,
    Pdf,
    Markdown,
    Plain,
    Json,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum KnowledgeNodeType {
    Concept,
    Person,
    Organization,
    Company,
    Paper,
    Book,
    Experiment,
    Event,
    Technology,
    Product,
    Application,
    Policy,
    Dataset,
    Controversy,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum NodeStatus {
    Active,
    Merged,
    Superseded,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum Confidence {
    Confirmed,
    High,
    Medium,
    Low,
    Unverified,
    Conflicting,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum ReviewState {
    Unreviewed,
    NeedsReview,
    Reviewed,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum EvidenceDirection {
    Support,
    Contradict,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum RelationConfidence {
    Confirmed,
    High,
    Medium,
    Low,
    Unverified,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum ArtifactKind {
    VaultExport,
    Report,
    GraphSnapshot,
    CoverageReport,
    Other,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct SourceQuality {
    pub authority: Option<QualityRating>,
    pub fitness: Option<QualityRating>,
    pub evaluated_at: Option<String>,
    pub notes: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Source {
    pub schema_version: SchemaVersionV1,
    pub source_id: String,
    pub project_id: String,
    pub url: String,
    pub title: String,
    pub kind: SourceKind,
    pub published_at: Option<String>,
    pub retrieved_at: String,
    pub languages: Option<Vec<String>>,
    pub dedup_key: String,
    pub quality: Option<SourceQuality>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SourceContent {
    pub schema_version: SchemaVersionV1,
    pub content_id: String,
    pub source_id: String,
    pub format: ContentFormat,
    pub content_hash: String,
    pub language: Option<String>,
    pub fetched_at: String,
    pub cache_ref: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct KnowledgeNode {
    pub schema_version: SchemaVersionV1,
    pub node_id: String,
    pub project_id: String,
    #[serde(rename = "type")]
    pub node_type: KnowledgeNodeType,
    pub title: String,
    pub aliases: Option<Vec<String>>,
    pub summary: Option<String>,
    pub status: Option<NodeStatus>,
    pub confidence: Confidence,
    pub source_ids: Option<Vec<String>>,
    pub claim_ids: Option<Vec<String>>,
    pub metadata: Option<serde_json::Value>,
    pub created_at: String,
    pub updated_at: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct ClaimProvenance {
    pub created_by: Option<CreatedBy>,
    pub prompt_id: Option<String>,
    pub prompt_version: Option<String>,
    pub task_id: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum CreatedBy {
    User,
    Ai,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Claim {
    pub schema_version: SchemaVersionV1,
    pub claim_id: String,
    pub project_id: String,
    pub subject_node_id: String,
    pub predicate: String,
    pub object: String,
    pub scope: Option<String>,
    pub status: Confidence,
    pub evidence_ids: Option<Vec<String>>,
    pub provenance: Option<ClaimProvenance>,
    pub review_state: Option<ReviewState>,
    pub created_at: String,
    pub updated_at: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct EvidenceLocator {
    pub quote: Option<String>,
    pub page: Option<String>,
    pub section: Option<String>,
    pub fragment: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Evidence {
    pub schema_version: SchemaVersionV1,
    pub evidence_id: String,
    pub claim_id: String,
    pub source_id: String,
    pub direction: EvidenceDirection,
    pub locator: EvidenceLocator,
    pub extraction_method: Option<CreatedBy>,
    pub retrieved_at: String,
    pub revised_at: Option<String>,
    pub superseded_by: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum RelationDirection {
    Directed,
    Undirected,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Relation {
    pub schema_version: SchemaVersionV1,
    pub relation_id: String,
    pub from_node_id: String,
    pub to_node_id: String,
    pub predicate: String,
    pub direction: Option<RelationDirection>,
    pub confidence: RelationConfidence,
    pub status: Option<NodeStatus>,
    pub source_ids: Option<Vec<String>>,
    pub claim_ids: Option<Vec<String>>,
    pub provenance: Option<ClaimProvenance>,
    pub created_at: String,
    pub updated_at: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum ArtifactFormat {
    Markdown,
    Json,
    Png,
    Svg,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Artifact {
    pub schema_version: SchemaVersionV1,
    pub artifact_id: String,
    pub project_id: String,
    pub kind: ArtifactKind,
    pub run_id: Option<String>,
    pub path: Option<String>,
    pub format: Option<ArtifactFormat>,
    pub note: Option<String>,
    pub created_at: String,
}

use crate::ContractValid;

impl ContractValid for Source {
    fn check_contract(&self) -> Result<(), String> {
        if self.source_id.is_empty() || self.project_id.is_empty() {
            return Err("source_id and project_id must not be empty".into());
        }
        if self.url.is_empty() || self.title.is_empty() || self.dedup_key.is_empty() {
            return Err("url, title, and dedup_key must not be empty".into());
        }
        Ok(())
    }
}

impl ContractValid for SourceContent {
    fn check_contract(&self) -> Result<(), String> {
        if self.content_id.is_empty() || self.source_id.is_empty() {
            return Err("content_id and source_id must not be empty".into());
        }
        if self.content_hash.len() < 8 {
            return Err("content_hash must be at least 8 characters".into());
        }
        Ok(())
    }
}

impl ContractValid for KnowledgeNode {
    fn check_contract(&self) -> Result<(), String> {
        if self.node_id.is_empty() || self.project_id.is_empty() {
            return Err("node_id and project_id must not be empty".into());
        }
        if self.title.is_empty() {
            return Err("title must not be empty".into());
        }
        Ok(())
    }
}

impl ContractValid for Claim {
    fn check_contract(&self) -> Result<(), String> {
        if self.claim_id.is_empty() || self.project_id.is_empty() || self.subject_node_id.is_empty()
        {
            return Err("claim_id, project_id, and subject_node_id must not be empty".into());
        }
        if self.predicate.is_empty() || self.object.is_empty() {
            return Err("predicate and object must not be empty".into());
        }
        Ok(())
    }
}

impl ContractValid for Evidence {
    fn check_contract(&self) -> Result<(), String> {
        if self.evidence_id.is_empty() || self.claim_id.is_empty() || self.source_id.is_empty() {
            return Err("evidence_id, claim_id, and source_id must not be empty".into());
        }
        Ok(())
    }
}

impl ContractValid for Relation {
    fn check_contract(&self) -> Result<(), String> {
        if self.relation_id.is_empty() || self.from_node_id.is_empty() || self.to_node_id.is_empty()
        {
            return Err("relation_id, from_node_id, and to_node_id must not be empty".into());
        }
        if self.predicate.is_empty() {
            return Err("predicate must not be empty".into());
        }
        Ok(())
    }
}

impl ContractValid for Artifact {
    fn check_contract(&self) -> Result<(), String> {
        if self.artifact_id.is_empty() || self.project_id.is_empty() {
            return Err("artifact_id and project_id must not be empty".into());
        }
        Ok(())
    }
}
